File: ITing-backend/src/main/java/com/iting/jobportal/admin/rbac/service/RbacService.java
package com.iting.jobportal.admin.rbac.service;

import com.iting.jobportal.admin.rbac.dto.*;
import java.util.List;

/** Quản lý RBAC: catalog quyền, vai trò (platform/company), vòng đời duyệt và gán role. */
public interface RbacService {

  List<PermissionResponse> listPermissions(String scope);

  List<RoleResponse> listRoles(String scope);

  List<RoleResponse> listPendingApprovals();

  RoleResponse getRole(Long id);

  RoleResponse createRole(Long actorId, CreateRoleRequest request);

  RoleResponse updateRole(Long actorId, Long id, UpdateRoleRequest request);

  /** DRAFT → PENDING_APPROVAL. */
  RoleResponse submitForApproval(Long actorId, Long id);

  /** PENDING_APPROVAL → ACTIVE. Chỉ Super Admin. */
  RoleResponse approveRole(Long actorId, Long id);

  /** PENDING_APPROVAL → REJECTED. Chỉ Super Admin. */
  RoleResponse rejectRole(Long actorId, Long id, String reason);

  /** Bật/tắt role đã duyệt: ACTIVE ⇄ DISABLED. */
  RoleResponse setStatus(Long actorId, Long id, String status);

  void deleteRole(Long actorId, Long id);

  /** Gán role cho tài khoản (kiểm tra account_type & các rule bảo mật). */
  void assignRole(Long actorId, AssignRoleRequest request);
}

File: ITing-frontend/src/services/rbacService.js
import axiosInstance from "../utils/axiosInstance";

// RBAC — vai trò nền tảng (PLATFORM) / doanh nghiệp (COMPANY) + quy trình duyệt.
// axiosInstance đã unwrap response.data nên các hàm trả thẳng dữ liệu.
const rbacService = {
  // Catalog quyền (scope: 'PLATFORM' | 'COMPANY' | undefined = tất cả)
  getPermissions: (scope) =>
    axiosInstance.get("/admin/rbac/permissions", { params: scope ? { scope } : {} }),

  // Vai trò
  getRoles: (scope) =>
    axiosInstance.get("/admin/rbac/roles", { params: scope ? { scope } : {} }),
  getPendingRoles: () => axiosInstance.get("/admin/rbac/roles/pending"),
  getRole: (id) => axiosInstance.get(`/admin/rbac/roles/${id}`),
  createRole: (payload) => axiosInstance.post("/admin/rbac/roles", payload),
  updateRole: (id, payload) => axiosInstance.put(`/admin/rbac/roles/${id}`, payload),
  deleteRole: (id) => axiosInstance.delete(`/admin/rbac/roles/${id}`),

  // Vòng đời duyệt
  submitRole: (id) => axiosInstance.post(`/admin/rbac/roles/${id}/submit`),
  approveRole: (id) => axiosInstance.post(`/admin/rbac/roles/${id}/approve`),
  rejectRole: (id, reason) =>
    axiosInstance.post(`/admin/rbac/roles/${id}/reject`, { reason }),
  setRoleStatus: (id, status) =>
    axiosInstance.patch(`/admin/rbac/roles/${id}/status`, { status }),

  // Gán role cho tài khoản
  assignRole: (accountId, roleCode) =>
    axiosInstance.post("/admin/rbac/assign", { accountId, roleCode }),

  // ── Company RBAC: HR con trong công ty ──────────────────────────────
  getCompanies: (keyword) =>
    axiosInstance.get("/admin/rbac/companies", { params: keyword ? { keyword } : {} }),
  getCompanyMembers: (companyId) =>
    axiosInstance.get(`/admin/rbac/companies/${companyId}/members`),
  getAvailableHr: (keyword) =>
    axiosInstance.get("/admin/rbac/companies/available-hr", { params: keyword ? { keyword } : {} }),
  addCompanyMember: (companyId, accountId, roleCode) =>
    axiosInstance.post(`/admin/rbac/companies/${companyId}/members`, { accountId, roleCode }),
  assignMemberRole: (affiliationId, roleCode) =>
    axiosInstance.put(`/admin/rbac/members/${affiliationId}/role`, { roleCode }),
  removeCompanyMember: (affiliationId) =>
    axiosInstance.delete(`/admin/rbac/members/${affiliationId}`),

  // ── Tài khoản nội bộ (gán platform role) ────────────────────────────
  getStaff: (keyword) =>
    axiosInstance.get("/admin/rbac/staff", { params: keyword ? { keyword } : {} }),
  // Tạo tài khoản nội bộ mới (đăng nhập được). payload: {email, password, fullName, platformRoleCode?}
  createStaff: (payload) => axiosInstance.post("/admin/rbac/staff", payload),
  promoteStaff: (accountId) =>
    axiosInstance.post(`/admin/rbac/staff/${accountId}/promote`),
  clearStaffRole: (accountId) =>
    axiosInstance.delete(`/admin/rbac/staff/${accountId}/role`),

  // Quyền hiệu lực của user hiện tại (ẩn/hiện menu & nút)
  getMyPermissions: () => axiosInstance.get("/admin/rbac/me/permissions"),
};

export default rbacService;
